### PR TITLE
39 simplify the table printer

### PR DESCRIPTION
**Pull request by els0r**
# Summary

Provides a `results.Result` structure that is returned to a printing routine. This data structure itself allows aggregation of its rows in case results from different systems should be aggregated.

It's the first step towards a distributed-query tool.

As preparation, it also adds support for `ctx` to the `query.Execute` and `query.Print` (new) methods, allowing callers to cancel any ongoing query operations. So far that was only possible via killing the command line tool. Thus, a critical flaw of the API is removed, where queries would continue even though the request was cancelled.

### CODE REVIEW

File: pkg/goDB/keyval.go [hunk 1]
```diff
@@ -68,9 +71,9 @@ func (k Key) MarshalJSON() ([]byte, error) {
 			Dport uint16 `json:"dport"`
 			Proto string `json:"ip_protocol"`
 		}{
-			RawIPToString(k.Sip[:]),
-			RawIPToString(k.Dip[:]),
-			uint16(uint16(k.Dport[0])<<8 | uint16(k.Dport[1])),
```
Comment (fako1024) [inline pkg/goDB/keyval.go:73]: The outer uint16() should be unnecessary (doesn't your linter make mimimi about that?)

File: cmd/goQuery/commands/help.go [hunk 2]
```diff
@@ -2,7 +2,7 @@ package commands
 
 var helpBase = `
   goquery -i <interfaces> [-hax] [--in|--out|--sum] [-n <max_n>] [--resolve]
-  [-e txt|csv|json|influxdb] [-d <db-path>] [-f <timestamp>] [-l <timestamp>]
```
Comment (fako1024) [inline cmd/goQuery/commands/help.go:5]: The horror! ;-)

File: pkg/goDB/DBWorkManager.go [hunk 3]
```diff
@@ -288,88 +246,93 @@ func (w *DBWorkManager) readBlocksAndEvaluate(workload DBWorkload, resultMap map
...
-			// Check whether conditional is satisfied for current entry
-			var conditionalSatisfied bool
-			if query.Conditional == nil {
-				conditionalSatisfied = true
-			} else {
-				// Populate comparison value for current entry
-				for _, colIdx := range query.conditionalAttributeIndizes {
-					copyToKeyFns[colIdx](i, &comparisonValue, blocks[colIdx])
+			// Check whether all blocks have matching number of entries
+			numEntries := int(len(blocks[BytesRcvdColIdx]) / 8)
```
Comment (fako1024) [inline pkg/goDB/DBWorkManager.go:278]: This will create conflicts with #37 (needs careful cross-checking to avoid unnoticed issues)...

**Review by fako1024 — CHANGES_REQUESTED** [review]
Nice one, some neat cleanup going on there! Just some very minor feedback, otherwise lgtm!

**Comment by fako1024** [comment]
Ran a few queries, seems to yield the same results as before from what I gather... :+1:

File: pkg/goDB/Query.go [hunk 4]
```diff
@@ -124,24 +126,34 @@ func NewQuery(attributes []Attribute, conditional Node, hasAttrTime, hasAttrIfac
 
 	for _, attrib := range q.Attributes {
 		colIdx := queryAttributeNameToColumnIndex(attrib.Name())
-		q.queryAttributeIndizes = append(q.queryAttributeIndizes, colIdx)
+		q.queryAttributeIndices = append(q.queryAttributeIndices, colIdx)
```
[the remark this answers is not in the document]
Comment (els0r) [reply pkg/goDB/Query.go:129]: Freaked out every time I saw it :D

File: pkg/goDB/DBWorkManager.go [hunk 5]
```diff
@@ -288,88 +246,93 @@ func (w *DBWorkManager) readBlocksAndEvaluate(workload DBWorkload, resultMap map
...
+				blocks      [ColIdxCount][]byte
+				blockBroken bool
+			)
+
+			for _, colIdx := range query.columnIndices {
+
+				// Read the block from the file
+				if blocks[colIdx], err = columnFiles[colIdx].ReadBlock(tstamp); err != nil {
+					blockBroken = true
+					w.logger.Warnf("[D %s; B %d] Failed to read %s.gpf: %s", dir, tstamp, columnFileNames[colIdx], err.Error())
```
[the remark this answers is not in the document]
Comment (els0r) [reply pkg/goDB/DBWorkManager.go:264]: Got it. Will factor this in for the logging refactor. `goQuery` will then use a global logger that logs to `os.Stderr` by default unless someone really wants to log to a file.

Fully agree that having to guess whether something went wrong vs. just seeing it, is suboptimal.

**Review by fako1024 — APPROVED** [review]

**Pull request merged**